stuffer/s2n_stuffer_base64.c

```diff
@@ -13,124 +13,58 @@
-    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-    255, 255, 255, 255, 255, 62, 255, 255, 255, 63, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 255, 255, 255, 64, 255, 255,
-    255, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 255, 255, 255,
-    255, 255, 255, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
-    51, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
-    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255 };
-
 bool s2n_is_base64_char(unsigned char c)
 {
-    return (b64_inverse[*((uint8_t *) (&c))] != 255);
+    /* use bitwise operations to minimize branching */
+    uint8_t out = 0;
+    out ^= (c >= 'A') & (c <= 'Z');
+    out ^= (c >= 'a') & (c <= 'z');
+    out ^= (c >= '0') & (c <= '9');
+    out ^= c == '+';
+    out ^= c == '/';
+    out ^= c == '=';
+
+    return out == 1;
 }
 
-/**
- * NOTE:
- * In general, shift before masking. This avoids needing to worry about how the
- * signed bit may be handled.
- */
 int s2n_stuffer_read_base64(struct s2n_stuffer *stuffer, struct s2n_stuffer *out)
 {
     POSIX_PRECONDITION(s2n_stuffer_validate(stuffer));
     POSIX_PRECONDITION(s2n_stuffer_validate(out));
-    int bytes_this_round = 3;
-    s2n_stack_blob(o, 4, 4);
 
-    do {
-        if (s2n_stuffer_data_available(stuffer) < o.size) {
-            break;
+    int base64_data_size = s2n_stuffer_data_available(stuffer) / 4 * 4;
+    int binary_output_size = base64_data_size / 4 * 3;
+
+    POSIX_GUARD(s2n_stuffer_skip_read(stuffer, base64_data_size));
+    const uint8_t *start_of_base64_data = stuffer->blob.data + stuffer->read_cursor - base64_data_size;
+
+    POSIX_GUARD(s2n_stuffer_skip_write(out, binary_output_size));
+    uint8_t *start_of_binary_output = out->blob.data + out->write_cursor - binary_output_size;
+
+    int res = EVP_DecodeBlock(start_of_binary_output, start_of_base64_data, base64_data_size);
+    POSIX_ENSURE(res > 0, S2N_ERR_INVALID_BASE64);
+
+    /* > The output will be padded with 0 bits if necessary to ensure that the 
+     * > output is always 3 bytes for every 4 input bytes. This function will 
+     * > return the length of the data decoded or -1 on error.
+     * https://docs.openssl.org/1.1.1/man3/EVP_EncodeInit/
```

Comment:
In the s2n-tls codebase, the quote comes after the source, even for legacy comments that aren't actually duvet citations.

Also, did you mean to include the comment about return values, or should that be on the previous return check code?